File: typescript/tuple.ts
export class Tuple<T> {
    private _items: T[];

    constructor(...items: T[]) {
        this._items = [...items];
    }

    get size(): number {
        return this._items.length;
    }

    get isEmpty(): boolean {
        return this._items.length === 0;
    }

    at(index: number): T | undefined {
        if (index < 0 || index >= this._items.length) {
            return undefined;
        }
        return this._items[index];
    }

    first(): T | undefined {
        return this._items[0];
    }

    last(): T | undefined {
        return this._items[this._items.length - 1];
    }

    toArray(): T[] {
        return [...this._items];
    }

    map<U>(func: (item: T) => U): Tuple<U> {
        return new Tuple(...this._items.map(func));
    }

    filter(func: (item: T) => boolean): Tuple<T> {
        return new Tuple(...this._items.filter(func));
    }

    reduce<U>(func: (accumulator: U, current: T) => U, initial: U): U {
        return this._items.reduce(func, initial);
    }

    concat(other: Tuple<T>): Tuple<T> {
        return new Tuple(...this._items, ...other._items);
    }

    slice(start: number, end?: number): Tuple<T> {
        return new Tuple(...this._items.slice(start, end));
    }

    take(n: number): Tuple<T> {
        return this.slice(0, n);
    }

    drop(n: number): Tuple<T> {
        return this.slice(n);
    }

    contains(item: T, equal?: (a: T, b: T) => boolean): boolean {
        const eq = equal || ((a, b) => a === b);
        return this._items.some(i => eq(i, item));
    }

    indexOf(item: T, equal?: (a: T, b: T) => boolean): number {
        const eq = equal || ((a, b) => a === b);
        for (let i = 0; i < this._items.length; i++) {
            if (eq(this._items[i], item)) {
                return i;
            }
        }
        return -1;
    }

    count(item: T, equal?: (a: T, b: T) => boolean): number {
        const eq = equal || ((a, b) => a === b);
        return this._items.filter(i => eq(i, item)).length;
    }

    reverse(): Tuple<T> {
        return new Tuple(...this._items.reverse());
    }

    [Symbol.iterator](): Iterator<T> {
        return this._items[Symbol.iterator]();
    }

    toString(): string {
        return `Tuple(${this._items.join(', ')})`;
    }
}

export class Pair<T, U> {
    private _first: T;
    private _second: U;

    constructor(first: T, second: U) {
        this._first = first;
        this._second = second;
    }

    get first(): T {
        return this._first;
    }

    get second(): U {
        return this._second;
    }

    swap(): Pair<U, T> {
        return new Pair(this._second, this._first);
    }

    toTuple(): Tuple<any> {
        return new Tuple<any>(this._first, this._second);
    }

    mapFirst<V>(func: (item: T) => V): Pair<V, U> {
        return new Pair(func(this._first), this._second);
    }

    mapSecond<V>(func: (item: U) => V): Pair<T, V> {
        return new Pair(this._first, func(this._second));
    }

    mapBoth<V, W>(func1: (item: T) => V, func2: (item: U) => W): Pair<V, W> {
        return new Pair(func1(this._first), func2(this._second));
    }

    toString(): string {
        return `Pair(${this._first}, ${this._second})`;
    }
}

export class Triple<T, U, V> {
    private _first: T;
    private _second: U;
    private _third: V;

    constructor(first: T, second: U, third: V) {
        this._first = first;
        this._second = second;
        this._third = third;
    }

    get first(): T {
        return this._first;
    }

    get second(): U {
        return this._second;
    }

    get third(): V {
        return this._third;
    }

    toTuple(): Tuple<any> {
        return new Tuple<any>(this._first, this._second, this._third);
    }

    mapFirst<W>(func: (item: T) => W): Triple<W, U, V> {
        return new Triple(func(this._first), this._second, this._third);
    }

    mapSecond<W>(func: (item: U) => W): Triple<T, W, V> {
        return new Triple(this._first, func(this._second), this._third);
    }

    mapThird<W>(func: (item: V) => W): Triple<T, U, W> {
        return new Triple(this._first, this._second, func(this._third));
    }

    toString(): string {
        return `Triple(${this._first}, ${this._second}, ${this._third})`;
    }
}

export function zip<T, U>(t1: Tuple<T>, t2: Tuple<U>): Tuple<Pair<T, U>> {
    const minLength = Math.min(t1.size, t2.size);
    const result: Pair<T, U>[] = [];
    for (let i = 0; i < minLength; i++) {
        const item1 = t1.at(i)!;
        const item2 = t2.at(i)!;
        result.push(new Pair(item1, item2));
    }
    return new Tuple(...result);
}

// if (require.main === module) {
//     console.log("=== Tuple Example ===");
//     const t1 = new Tuple(1, 2, 3, 4, 5);
//     console.log(`Tuple: ${t1}`);
//     console.log(`Size: ${t1.size}`);
//     console.log(`First: ${t1.first()}`);
//     console.log(`Last: ${t1.last()}`);
//     console.log(`At index 2: ${t1.at(2)}`);
//     console.log();

//     console.log("=== Tuple Operations ===");
//     const t2 = new Tuple("a", "b", "c");
//     console.log(`Tuple t2: ${t2}`);
//     console.log(`Concat t1 + t2: ${t1.concat(t2 as any)}`);
//     console.log(`Slice t1[1:4]: ${t1.slice(1, 4)}`);
//     console.log(`Take 3 from t1: ${t1.take(3)}`);
//     console.log(`Drop 2 from t1: ${t1.drop(2)}`);
//     console.log(`Reverse t1: ${t1.reverse()}`);
//     console.log(`Map t1 (x * 2): ${t1.map(x => x * 2)}`);
//     console.log(`Filter t1 (even): ${t1.filter(x => x % 2 === 0)}`);
//     console.log(`Reduce t1 (sum): ${t1.reduce((a, b) => a + b, 0)}`);
//     console.log();

//     console.log("=== Pair Example ===");
//     const p = new Pair(10, "hello");
//     console.log(`Pair: ${p}`);
//     console.log(`First: ${p.first}`);
//     console.log(`Second: ${p.second}`);
//     console.log(`Swap: ${p.swap()}`);
//     console.log(`Map first (+5): ${p.mapFirst(x => x + 5)}`);
//     console.log(`Map second (upper): ${p.mapSecond(s => s.toUpperCase())}`);
//     console.log();

//     console.log("=== Triple Example ===");
//     const tri = new Triple("a", 100, true);
//     console.log(`Triple: ${tri}`);
//     console.log(`First: ${tri.first}`);
//     console.log(`Second: ${tri.second}`);
//     console.log(`Third: ${tri.third}`);
//     console.log();

//     console.log("=== Zip Example ===");
//     const t3 = new Tuple(1, 2, 3);
//     const t4 = new Tuple("x", "y", "z");
//     const zipped = zip(t3, t4);
//     console.log(`Zip ${t3} and ${t4}: ${zipped}`);
//     console.log();

//     console.log("=== Comparison Example ===");
//     const t5 = new Tuple(1, 2, 3);
//     console.log(`Contains 2 in t5: ${t5.contains(2)}`);
//     console.log(`Index of 3 in t5: ${t5.indexOf(3)}`);
// }

File: rust/src/tuple.rs
#![allow(dead_code)]

use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Tuple<T: Clone> {
    items: Vec<T>,
}

impl<T: Clone> Tuple<T> {
    pub fn new(items: Vec<T>) -> Self {
        Tuple { items }
    }

    pub fn from_slice(items: &[T]) -> Self {
        Tuple {
            items: items.to_vec(),
        }
    }

    pub fn size(&self) -> usize {
        self.items.len()
    }

    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }

    pub fn at(&self, index: usize) -> Option<&T> {
        self.items.get(index)
    }

    pub fn first(&self) -> Option<&T> {
        self.items.first()
    }

    pub fn last(&self) -> Option<&T> {
        self.items.last()
    }

    pub fn to_vec(&self) -> Vec<T> {
        self.items.clone()
    }

    pub fn map<U: Clone, F: Fn(&T) -> U>(&self, f: F) -> Tuple<U> {
        Tuple {
            items: self.items.iter().map(f).collect(),
        }
    }

    pub fn filter<F: Fn(&T) -> bool>(&self, f: F) -> Tuple<T> {
        Tuple {
            items: self.items.iter().filter(|&x| f(x)).cloned().collect(),
        }
    }

    pub fn reduce<F: Fn(T, &T) -> T>(&self, f: F, initial: T) -> T {
        self.items.iter().fold(initial, f)
    }

    pub fn concat(&self, other: &Tuple<T>) -> Tuple<T> {
        let mut result = self.items.clone();
        result.extend(other.items.clone());
        Tuple { items: result }
    }

    pub fn slice(&self, start: usize, end: usize) -> Tuple<T> {
        let start = start.min(self.items.len());
        let end = end.min(self.items.len());
        if start >= end {
            return Tuple { items: vec![] };
        }
        Tuple {
            items: self.items[start..end].to_vec(),
        }
    }

    pub fn take(&self, n: usize) -> Tuple<T> {
        self.slice(0, n)
    }

    pub fn drop(&self, n: usize) -> Tuple<T> {
        self.slice(n, self.items.len())
    }

    pub fn contains<F: Fn(&T) -> bool>(&self, f: F) -> bool {
        self.items.iter().any(f)
    }

    pub fn find<F: Fn(&T) -> bool>(&self, f: F) -> Option<usize> {
        self.items.iter().position(f)
    }

    pub fn count<F: Fn(&T) -> bool>(&self, f: F) -> usize {
        self.items.iter().filter(|&x| f(x)).count()
    }

    pub fn reverse(&self) -> Tuple<T> {
        let mut items = self.items.clone();
        items.reverse();
        Tuple { items }
    }

    pub fn iter(&self) -> std::slice::Iter<'_, T> {
        self.items.iter()
    }
}

impl<T: Clone + fmt::Display> fmt::Display for Tuple<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let items: Vec<String> = self.items.iter().map(|x| x.to_string()).collect();
        write!(f, "Tuple({})", items.join(", "))
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Pair<T: Clone, U: Clone> {
    first: T,
    second: U,
}

impl<T: Clone, U: Clone> Pair<T, U> {
    pub fn new(first: T, second: U) -> Self {
        Pair { first, second }
    }

    pub fn first(&self) -> &T {
        &self.first
    }

    pub fn second(&self) -> &U {
        &self.second
    }

    pub fn swap(&self) -> Pair<U, T> {
        Pair::new(self.second.clone(), self.first.clone())
    }

    pub fn to_tuple(&self) -> (T, U) {
        (self.first.clone(), self.second.clone())
    }

    pub fn map_first<F: Fn(T) -> T>(&self, f: F) -> Pair<T, U> {
        Pair::new(f(self.first.clone()), self.second.clone())
    }

    pub fn map_second<F: Fn(U) -> U>(&self, f: F) -> Pair<T, U> {
        Pair::new(self.first.clone(), f(self.second.clone()))
    }

    pub fn map_both<F1: Fn(T) -> T, F2: Fn(U) -> U>(
        &self,
        f1: F1,
        f2: F2,
    ) -> Pair<T, U> {
        Pair::new(f1(self.first.clone()), f2(self.second.clone()))
    }
}

impl<T: Clone + fmt::Display, U: Clone + fmt::Display> fmt::Display for Pair<T, U> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Pair({}, {})", self.first, self.second)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Triple<T: Clone, U: Clone, V: Clone> {
    first: T,
    second: U,
    third: V,
}

impl<T: Clone, U: Clone, V: Clone> Triple<T, U, V> {
    pub fn new(first: T, second: U, third: V) -> Self {
        Triple {
            first,
            second,
            third,
        }
    }

    pub fn first(&self) -> &T {
        &self.first
    }

    pub fn second(&self) -> &U {
        &self.second
    }

    pub fn third(&self) -> &V {
        &self.third
    }

    pub fn map_first<F: Fn(&T) -> T>(&self, f: F) -> Triple<T, U, V> {
        Triple::new(f(&self.first), self.second.clone(), self.third.clone())
    }

    pub fn map_second<F: Fn(&U) -> U>(&self, f: F) -> Triple<T, U, V> {
        Triple::new(self.first.clone(), f(&self.second), self.third.clone())
    }

    pub fn map_third<F: Fn(&V) -> V>(&self, f: F) -> Triple<T, U, V> {
        Triple::new(self.first.clone(), self.second.clone(), f(&self.third))
    }
}

impl<T: Clone + fmt::Display, U: Clone + fmt::Display, V: Clone + fmt::Display> fmt::Display
    for Triple<T, U, V>
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Triple({}, {}, {})",
            self.first, self.second, self.third
        )
    }
}

pub fn zip<T: Clone, U: Clone>(t1: &Tuple<T>, t2: &Tuple<U>) -> Tuple<Pair<T, U>> {
    let min_len = t1.size().min(t2.size());
    let mut result = Vec::with_capacity(min_len);
    for i in 0..min_len {
        result.push(Pair::new(
            t1.at(i).unwrap().clone(),
            t2.at(i).unwrap().clone(),
        ));
    }
    Tuple::new(result)
}

pub fn run_tuple_examples() {
    println!("=== Tuple Example ===");
    let t1 = Tuple::new(vec![1, 2, 3, 4, 5]);
    println!("Tuple: {}", t1);
    println!("Size: {}", t1.size());
    println!("First: {:?}", t1.first());
    println!("Last: {:?}", t1.last());
    println!("At index 2: {:?}", t1.at(2));
    println!();

    println!("=== Tuple Operations ===");
    let t2 = Tuple::new(vec!["a", "b", "c"]);
    println!("Tuple t2: {}", t2);
    println!("Concat t1 + t2: {}", t1.concat(&Tuple::new(vec![])));
    println!("Slice t1[1..4]: {}", t1.slice(1, 4));
    println!("Take 3 from t1: {}", t1.take(3));
    println!("Drop 2 from t1: {}", t1.drop(2));
    println!("Reverse t1: {}", t1.reverse());
    println!("Map t1 (x * 2): {}", t1.map(|&x| x * 2));
    println!("Filter t1 (even): {}", t1.filter(|&x| x % 2 == 0));
    println!(
        "Reduce t1 (sum): {}",
        t1.reduce(|a, &b| a + b, 0)
    );
    println!();

    println!("=== Pair Example ===");
    let p = Pair::new(10, "hello");
    println!("Pair: {}", p);
    println!("First: {}", p.first());
    println!("Second: {}", p.second());
    println!("Swap: {}", p.swap());
    println!("Map first (+5): {}", p.map_first(|x| x + 5));
    println!();

    println!("=== Triple Example ===");
    let tri = Triple::new("a", 100, true);
    println!("Triple: {}", tri);
    println!("First: {}", tri.first());
    println!("Second: {}", tri.second());
    println!("Third: {}", tri.third());
    println!();

    println!("=== Zip Example ===");
    let t3 = Tuple::new(vec![1, 2, 3]);
    let t4 = Tuple::new(vec!["x", "y", "z"]);
    let zipped = zip(&t3, &t4);
    print!("Zip {} and {}: Tuple(", t3, t4);
    for (i, pair) in zipped.iter().enumerate() {
        if i > 0 {
            print!(", ");
        }
        print!("({}, {})", pair.first(), pair.second());
    }
    println!(")");
    println!();

    println!("=== Comparison Example ===");
    let t5 = Tuple::new(vec![1, 2, 3]);
    println!("Contains 2 in t5: {}", t5.contains(|&x| x == 2));
    println!("Index of 3 in t5: {:?}", t5.find(|&x| x == 3));
}

fn main() {
    run_tuple_examples();
}
